reindexer/complete_reindex/src/complete_reindex.py

```diff
@@ -2,31 +2,80 @@
 
 import os
 
+from boto3.dynamodb.conditions import Attr
 import boto3
-from botocore.exceptions import ClientError
+
 from wellcome_aws_utils import sns_utils
 
 
-def main(event, _ctxt=None, dynamodb_client=None):
-    print(f'event = {event!r}')
-    dynamodb_client = dynamodb_client or boto3.client('dynamodb')
+def _get_item_version(dynamo_item):
+    dynamo_item_version = 1
+
+    if 'version' in dynamo_item:
+        dynamo_item_version = dynamo_item['version']
+
+    return dynamo_item_version
+
+
+def _update_versioned_item(table, dynamo_item, desired_item):
+    desired_dynamo_conditional_update_version = (
+            _get_item_version(dynamo_item) + 1
+    )
+
+    new_dynamo_item = {**dynamo_item, **desired_item}
+    new_dynamo_item['version'] = desired_dynamo_conditional_update_version
+
+    print(f'Attempting conditional update to {new_dynamo_item}')
+
+    table.put_item(
+        Item=new_dynamo_item,
+        ConditionExpression=Attr('version').not_exists() | Attr('version').lt(
+            desired_dynamo_conditional_update_version
+        )
+    )
+
+
+def _process_reindex_tracker_update_job(table, message):
+    shard_id = message['shardId']
+    completed_reindex_version = message['completedReindexVersion']
+
+    dynamodb_response = table.get_item(Key={'shardId': shard_id})
 
-    table_name = os.environ['TABLE_NAME']
+    dynamo_item = dynamodb_response['Item']
 
+    print(f'Retrieved {dynamo_item}')
+
+    dynamo_current_version = dynamo_item['currentVersion']
+
+    if (dynamo_current_version >= completed_reindex_version):
+        print(f'Update for {shard_id} discarded as current version advanced.')
+        return
+
+    desired_item = {
+        "shardId": shard_id,
+        "desiredVersion": completed_reindex_version,
+        "currentVersion": completed_reindex_version,
+    }
+
+    return {
+        "dynamo_item": dynamo_item,
+        "desired_item": desired_item
+    }
+
+
+def _run(table, event):
     for record in sns_utils.extract_sns_messages_from_lambda_event(event):
-        row = record.message
-        try:
-            dynamodb_client.update_item(
-                TableName=table_name,
-                Key={'shardId': {'S': row['shardId']}},
-                UpdateExpression='SET currentVersion=:completedReindexVersion',
-                ConditionExpression='currentVersion < :completedReindexVersion',
-                ExpressionAttributeValues={
-                    ':completedReindexVersion': {'N': str(row['completedReindexVersion'])},
-                }
-            )
-        except ClientError as err:
-            if err.response['Error']['Code'] == 'ConditionalCheckFailedException':
-                print(f'{row["shardId"]} already has a newer currentVersion')
-            else:
-                raise
+        job = _process_reindex_tracker_update_job(table, record.message)
+
+    if job is not None:
```

Comment:
Should this be in the body of the loop? – you construct a job for every update, but only process the last one. Possibly needs extra tests?